modules/unsupported/vector-mosaic/src/main/java/org/geotools/vectormosaic/VectorMosaicFeatureSource.java

```diff
@@ -0,0 +1,340 @@
+                        }
+                        allBounds.add(env);
+                    }
+                }
+            } catch (Exception e) {
+                String message = "Failed to get the bounds on " + granule.getName();
+                if (getStore().isTolerant()) {
+                    LOGGER.log(Level.WARNING, message, e);
+                    return null;
+                } else {
+                    throw new IOException(message, e);
+                }
+            }
+        }
+        ReferencedEnvelope result =
+                new ReferencedEnvelope(getSchema().getCoordinateReferenceSystem());
+        for (ReferencedEnvelope bound : allBounds) {
+            Envelope env = new Envelope(bound);
+            result.expandToInclude(env);
+        }
+        if (result.isNull()) {
+            return null;
+        } else {
+            return result;
+        }
+    }
+
+    @Override
+    protected int getCountInternal(Query query) throws IOException {
+        List<Long> counts = new ArrayList<>();
+        for (VectorMosaicGranule granule : config.getGranules()) {
+            DataStore dataStore = granule.getDataStore();
+            if (dataStore != null) {
+                SimpleFeatureSource source =
+                        dataStore.getFeatureSource(granule.getGranuleTypeName());
+                Query q = new Query(query);
+                q.setTypeName(granule.getGranuleTypeName());
+                Long count = Long.valueOf(source.getCount(q));
+                counts.add(count);
+            }
+        }
+        // aggregate the counts
+        long total = 0;
+        for (Long count : counts) {
+            try {
+                if (count > 0) {
+                    total += count;
+                } else {
+                    // one of the sources found it was too costly to count
+                    return -1;
+                }
+            } catch (Exception e) {
+                throw new IOException("Failed to count on a delegate store", e);
+            }
+        }
+        return (int) total;
+    }
+
+    @Override
+    protected FeatureReader<SimpleFeatureType, SimpleFeature> getReaderInternal(Query query)
+            throws IOException {
+        try {
+            if (config.getGranules() == null || config.getGranules().isEmpty()) {
+                throw new IOException("No granules found for mosaic");
+            }
+            // get the first store
+            VectorMosaicGranule firstGranule = config.getGranules().get(0);
+            Query psQuery = new Query(query);
+            psQuery.setTypeName(firstGranule.getGranuleTypeName());
+            DataStore ps = firstGranule.getDataStore();
+            SimpleFeatureType target = DataUtilities.createView(ps, psQuery).getSchema();
+            target = retypeNameSchema(target);
+
+            FeatureQueue queue = new FeatureQueue();
+            for (VectorMosaicGranule vectorMosaicGranule : config.getGranules()) {
```

Comment:
Same as above. We'll have thousands of granules, we cannot go and query each one of them...

Reply:
Fixed